types/src/validator_verifier.rs

```diff
@@ -79,77 +84,96 @@ impl ValidatorConsensusInfo {
 /// Supports validation of signatures for known authors with individual voting powers. This struct
 /// can be used for all signature verification operations including block and network signature
 /// verification, respectively.
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
 pub struct ValidatorVerifier {
-    /// An ordered map of each validator's on-chain account address to its pubkeys
-    /// and voting power.
-    address_to_validator_info: BTreeMap<AccountAddress, ValidatorConsensusInfo>,
+    /// A vector of each validator's on-chain account address to its pubkeys and voting power.
+    validator_infos: Vec<ValidatorConsensusInfo>,
     /// The minimum voting power required to achieve a quorum
-    quorum_voting_power: u64,
+    #[serde(skip)]
+    quorum_voting_power: u128,
     /// Total voting power of all validators (cached from address_to_validator_info)
-    total_voting_power: u64,
+    #[serde(skip)]
+    total_voting_power: u128,
+    /// In-memory index of account address to its index in the vector, does not go through serde.
+    #[serde(skip)]
+    address_to_validator_index: HashMap<AccountAddress, usize>,
+}
+
+/// Reconstruct fields from the raw data upon deserialization.
+impl<'de> Deserialize<'de> for ValidatorVerifier {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(rename = "ValidatorVerifier")]
+        struct RawValidatorVerifier {
+            validator_infos: Vec<ValidatorConsensusInfo>,
+        }
+
+        let RawValidatorVerifier { validator_infos } =
+            RawValidatorVerifier::deserialize(deserializer)?;
+
+        Ok(ValidatorVerifier::new(validator_infos))
+    }
 }
 
 impl ValidatorVerifier {
+    /// Private constructor to calculate the in-memory index
+    fn _new(
```

Comment:
It's probably not a common convention to name functions with `_` to reflect private functions. How about we rename this to `new_with_voting_power` ?